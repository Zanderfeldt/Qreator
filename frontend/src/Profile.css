/* SignUpForm.css */

.profile {
  background-color: #ffffff62 ;
  border-radius: 10px;
  border: solid rgba(39, 35, 33, 0.192) 0.5px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.295);
  width: 320px;
  margin: 80px auto;
  padding: 20px;
  text-align: center;
  
}

.profile h2 {
  color: #000000;
  font-size: 24px;
  margin-bottom: 10px;
  margin-top: 5px;
}

.profile input {
  width: 80%;
  padding: 2px;
  margin: 4px 0;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: 14px;
  height: 25px;
  font-family: 'Mooli', sans-serif;
}

.profile label {
  display: block; 
  margin-top: 1px;
  font-size: 12px;
  font-weight: bold; 
  text-align: center;
  color: #25571e; 
}

.profile button {
  background-color: #62865d;
  color: #fff;
  padding: 5px 10px;
  border: none;
  border-radius: 1px;
  cursor: pointer;
  font-size: 16px;
  margin-top: 10px;
  font-family: 'Mooli', sans-serif;
}

.profile button:hover {
  background-color: #70dd5a;
}
