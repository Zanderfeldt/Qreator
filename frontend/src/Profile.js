import React, { useContext, useState } from 'react';
import './Profile.css';
import QreatorApi from './API';
import UserContext from './auth/UserContext';


function Profile() {
  const { currUser, setCurrUser } = useContext(UserContext);
  const INITIAL_STATE = {
    username: currUser.username,
    firstName: currUser.firstName,
    lastName: currUser.lastName,
    email: currUser.email,
    password: '',
  }

  const [formData, setFormData] = useState(INITIAL_STATE);
  const [showAlert, setShowAlert] = useState(false);
  const [formErrors, setFormErrors] = useState([]);
  

  const handleSubmit = async (e) => {
    e.preventDefault();

    let profileData = {
      username: formData.username,
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      password: formData.password,
    };

    
    let updatedUser;

    try {
      updatedUser = await QreatorApi.saveProfile(currUser.id, profileData);
    } catch (e) {
      setFormErrors(e);
      console.log(formErrors);
      return;
    }

    setFormData(f => ({ ...f, password: "" }));
    setFormErrors([]);

    // update currentUser with new profile information
    setCurrUser(updatedUser);
    setShowAlert(true);
    setTimeout(() => {
      setShowAlert(false);
    }, 3000);
  };

  const handleChange = e => {
    const {name, value} = e.target;
    setFormData(fData => ({
      ...fData,
      [name]: value
    }));
    setFormErrors([]);
  };

  const isFormValid = Object.values(formData).every(Boolean);

  return (
    <div className='profile'>
      <h2>Profile</h2>
      {showAlert && <div style={{color: 'green'}}>Changes saved successfully</div>}
      <form onSubmit={handleSubmit}>
          <label htmlFor='firstName'>- First Name -</label>
          <input
          id='firstName'
          name='firstName'
          placeholder='First Name'
          value={formData.firstName}
          onChange={handleChange}
          />
          <label htmlFor='lastName'>- Last Name -</label>
          <input
          id='lastName'
          name='lastName'
          placeholder='Last Name'
          value={formData.lastName}
          onChange={handleChange}
          />
          <label htmlFor='email'>- E-mail -</label>
          <input
          id='email'
          name='email'
          placeholder='E-mail'
          value={formData.email}
          onChange={handleChange}
          />
          <label htmlFor='username'>- Username -</label>
          <input
          id='username'
          name='username'
          placeholder='Username'
          value={formData.username}
          onChange={handleChange}
          />
          <label htmlFor='Password'>- Password (Required) -</label>
          <input
          id='password'
          type='password'
          name='password'
          placeholder='Password'
          value={formData.password}
          onChange={handleChange}
          />
          <button disabled={!isFormValid}>Save Changes</button>
      </form>

    </div>
  )
}

export default Profile;