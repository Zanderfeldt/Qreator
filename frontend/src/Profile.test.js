import React from 'react';
import { render } from '@testing-library/react';
import Profile from './Profile';
import UserContext from './auth/UserContext';
import { MemoryRouter } from 'react-router-dom'; 

const contextValue = {
  currUser: {
    firstName: 'test',
    lastName: 'test',
    username: 'test',
    email: 'test@test.com'
  }
};

//smoke test
test('renders Profile component without errors', () => {
  render(
  <MemoryRouter>
    <UserContext.Provider value={contextValue}>
      <Profile />
    </UserContext.Provider>
  </MemoryRouter>
  );
});

//snapshot
test('matches snapshot', () => {
  const { asFragment } = render(
  <MemoryRouter>
    <UserContext.Provider value={contextValue}>
      <Profile />
    </UserContext.Provider>
  </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

//test that input fields are pre-populated with currUser info
test('populates input fields with currUser values', () => {
  const { getByLabelText } = render(
    <MemoryRouter>
      <UserContext.Provider value={contextValue}>
        <Profile />
      </UserContext.Provider>
    </MemoryRouter>
  );

  // Get input fields by their associated label text
  const firstNameInput = getByLabelText('- First Name -');
  const lastNameInput = getByLabelText('- Last Name -');
  const usernameInput = getByLabelText('- Username -');
  const emailInput = getByLabelText('- E-mail -');

  // Assert that input values match the corresponding contextValue properties
  expect(firstNameInput).toHaveValue(contextValue.currUser.firstName);
  expect(lastNameInput).toHaveValue(contextValue.currUser.lastName);
  expect(usernameInput).toHaveValue(contextValue.currUser.username);
  expect(emailInput).toHaveValue(contextValue.currUser.email);
});