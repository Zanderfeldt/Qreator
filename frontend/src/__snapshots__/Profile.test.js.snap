// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="profile"
  >
    <h2>
      Profile
    </h2>
    <form>
      <label
        for="firstName"
      >
        - First Name -
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="First Name"
        value="test"
      />
      <label
        for="lastName"
      >
        - Last Name -
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        value="test"
      />
      <label
        for="email"
      >
        - E-mail -
      </label>
      <input
        id="email"
        name="email"
        placeholder="E-mail"
        value="test@test.com"
      />
      <label
        for="username"
      >
        - Username -
      </label>
      <input
        id="username"
        name="username"
        placeholder="Username"
        value="test"
      />
      <label
        for="Password"
      >
        - Password (Required) -
      </label>
      <input
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        disabled=""
      >
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
